File: content/develop/use-cases/prefetch-cache/go/sync_worker.go
// Background sync worker for the prefetch-cache demo.
//
// A daemon goroutine drains the primary's change channel and applies
// each event to Redis through PrefetchCache.ApplyChange. In a real
// system, the channel is replaced by a CDC pipeline (Redis Data
// Integration, Debezium, or an equivalent) that tails the primary's
// binlog/WAL and writes the same shape of events.
//
// The worker exposes Pause() and Resume() so maintenance paths
// (/reprefetch, Clear()) can stop event application without tearing
// the goroutine down. Pause() blocks until the worker is parked, so
// the caller knows no apply is in flight by the time it returns.
package prefetchcache

import (
	"context"
	"log"
	"sync"
	"time"
)

// SyncWorker drains primary change events into Redis on a goroutine.
type SyncWorker struct {
	primary       *MockPrimaryStore
	cache         *PrefetchCache
	pollTimeout   time.Duration

	mu          sync.Mutex
	running     bool
	cancel      context.CancelFunc
	done        chan struct{}
	paused      bool
	pausedIdle  chan struct{} // closed (replaced with a fresh chan) every time the loop parks
	resumeCh    chan struct{} // closed by Resume to wake the parked loop
}

// NewSyncWorker creates a SyncWorker. The default poll timeout (50 ms)
// matches the Python reference.
func NewSyncWorker(primary *MockPrimaryStore, cache *PrefetchCache) *SyncWorker {
	return &SyncWorker{
		primary:     primary,
		cache:       cache,
		pollTimeout: 50 * time.Millisecond,
		pausedIdle:  make(chan struct{}),
		resumeCh:    make(chan struct{}),
	}
}

// Start spawns the worker goroutine if it is not already running.
func (w *SyncWorker) Start() {
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.running {
		return
	}
	ctx, cancel := context.WithCancel(context.Background())
	done := make(chan struct{})
	w.cancel = cancel
	w.done = done
	w.running = true
	go w.run(ctx, done)
}

// Stop signals the worker to exit and waits up to joinTimeout for the
// goroutine to finish.
//
// If the join times out the worker is wedged inside ApplyChange; we
// leave w.running true so a subsequent Start() does not spawn a second
// worker on top of the orphan.
func (w *SyncWorker) Stop(joinTimeout time.Duration) {
	w.mu.Lock()
	if !w.running {
		w.mu.Unlock()
		return
	}
	cancel := w.cancel
	done := w.done
	// Close resumeCh inside the lock so a concurrent Resume cannot
	// pass the "already closed?" check and then race us to close()
	// the same channel twice (which would panic).
	closeOnce(w.resumeCh)
	w.mu.Unlock()

	cancel()

	select {
	case <-done:
		w.mu.Lock()
		w.running = false
		w.cancel = nil
		w.done = nil
		w.mu.Unlock()
	case <-time.After(joinTimeout):
		// Worker is wedged: leave running=true so Start() is a no-op
		// rather than producing a second worker.
	}
}

// Pause sets the pause flag and blocks until the worker confirms it is
// parked, up to timeout. Returns true if confirmed paused.
//
// While paused, change events accumulate on the primary's channel and
// apply in order after Resume(). Calling Pause while already paused is
// idempotent and returns immediately.
func (w *SyncWorker) Pause(timeout time.Duration) bool {
	w.mu.Lock()
	if !w.running {
		w.paused = true
		w.mu.Unlock()
		return true
	}
	if w.paused {
		idle := w.pausedIdle
		w.mu.Unlock()
		// Already paused -- wait for the current idle signal (which
		// is closed once the worker is parked).
		select {
		case <-idle:
			return true
		case <-time.After(timeout):
			return false
		}
	}
	// Replace the resume channel with a fresh one so any prior
	// Resume() does not immediately unblock this pause.
	w.resumeCh = make(chan struct{})
	// Reset the idle channel: a fresh one will be closed by the
	// worker when it parks.
	w.pausedIdle = make(chan struct{})
	idle := w.pausedIdle
	w.paused = true
	w.mu.Unlock()

	select {
	case <-idle:
		return true
	case <-time.After(timeout):
		return false
	}
}

// Resume clears the pause flag and wakes the parked worker goroutine.
func (w *SyncWorker) Resume() {
	w.mu.Lock()
	defer w.mu.Unlock()
	if !w.paused {
		return
	}
	w.paused = false
	// Close inside the lock so a concurrent Stop cannot pass the
	// "already closed?" check and then race us to close() the same
	// channel twice (which would panic).
	closeOnce(w.resumeCh)
}

// closeOnce closes ch if it isn't already closed. Callers MUST hold
// w.mu while invoking it (the non-blocking receive + close pair is not
// atomic on its own; the mutex provides the missing serialisation).
func closeOnce(ch chan struct{}) {
	select {
	case <-ch:
		// Already closed.
	default:
		close(ch)
	}
}

func (w *SyncWorker) run(ctx context.Context, done chan struct{}) {
	defer close(done)
	for {
		// Bail out promptly on cancel.
		select {
		case <-ctx.Done():
			return
		default:
		}

		// Park if paused.
		w.mu.Lock()
		paused := w.paused
		idle := w.pausedIdle
		resumeCh := w.resumeCh
		w.mu.Unlock()
		if paused {
			// Signal "I am parked, no apply in flight". Closing the
			// channel lets every waiter on Pause() observe it.
			select {
			case <-idle:
				// Already closed -- nothing to do.
			default:
				close(idle)
			}
			// Wait for Resume() to close resumeCh, or for Stop() to
			// cancel the context.
			select {
			case <-resumeCh:
			case <-ctx.Done():
				return
			}
			continue
		}

		change, ok := w.primary.NextChange(w.pollTimeout)
		if !ok {
			continue
		}
		if err := w.cache.ApplyChange(ctx, change); err != nil {
			// Demo behaviour: log and drop the event. A production
			// CDC consumer would retry with bounded backoff and
			// expose a dead-letter / error counter; see the guide's
			// "Production usage" section.
			log.Printf("[sync] failed to apply %s %s: %v", change.Op, change.ID, err)
		}
	}
}

File: content/develop/use-cases/prefetch-cache/java-jedis/SyncWorker.java
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Background sync worker for the prefetch-cache demo.
 *
 * <p>A daemon thread drains the primary's change queue and applies each
 * event to Redis through {@link PrefetchCache#applyChange(Map)}. In a
 * real system, the queue is replaced by a CDC pipeline (Redis Data
 * Integration, Debezium, or an equivalent) that tails the primary's
 * binlog/WAL and writes the same shape of events.</p>
 *
 * <p>The worker exposes {@link #pause()} and {@link #resume()} so
 * maintenance paths ({@code /reprefetch}, {@code clear()}) can stop
 * event application without tearing the thread down. {@code pause()}
 * blocks until the worker is parked, so the caller knows no apply is in
 * flight by the time it returns.</p>
 */
public class SyncWorker {

    private final MockPrimaryStore primary;
    private final PrefetchCache cache;
    private final long pollTimeoutMs;

    private final ReentrantLock lock = new ReentrantLock();
    /** Signalled when {@link #pause()} or {@link #resume()} changes the run-state flag. */
    private final Condition stateChanged = lock.newCondition();
    /** Signalled when the worker has confirmed it is parked. */
    private final Condition idleSignal = lock.newCondition();

    private volatile boolean stopRequested;
    private volatile boolean pauseRequested;
    private volatile boolean workerIdle;

    private Thread thread;

    public SyncWorker(MockPrimaryStore primary, PrefetchCache cache) {
        this(primary, cache, 50L);
    }

    public SyncWorker(MockPrimaryStore primary, PrefetchCache cache, long pollTimeoutMs) {
        if (primary == null || cache == null) {
            throw new IllegalArgumentException("primary and cache are required");
        }
        this.primary = primary;
        this.cache = cache;
        this.pollTimeoutMs = pollTimeoutMs;
    }

    /** Spawn the worker if it isn't already running. */
    public synchronized void start() {
        if (thread != null && thread.isAlive()) {
            return;
        }
        stopRequested = false;
        pauseRequested = false;
        workerIdle = false;
        thread = new Thread(this::run, "prefetch-cache-sync");
        thread.setDaemon(true);
        thread.start();
    }

    /**
     * Signal the worker to exit and join its thread.
     *
     * <p>If the join times out the worker is wedged inside an apply; we
     * leave {@code thread} populated so a subsequent {@link #start()}
     * does not spawn a second worker on top of the orphan.</p>
     */
    public synchronized void stop(long joinTimeoutMs) {
        stopRequested = true;
        lock.lock();
        try {
            stateChanged.signalAll();
        } finally {
            lock.unlock();
        }
        if (thread == null) {
            return;
        }
        try {
            thread.join(joinTimeoutMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        if (!thread.isAlive()) {
            thread = null;
        }
    }

    public void stop() {
        stop(2000L);
    }

    /**
     * Stop applying events and block until the worker is parked.
     *
     * <p>Returns {@code true} once the worker has confirmed it is idle,
     * or {@code false} if the timeout elapsed first. While paused,
     * change events accumulate in the primary's queue and are applied
     * in order after {@link #resume()}.</p>
     */
    public boolean pause(long timeoutMs) {
        lock.lock();
        try {
            pauseRequested = true;
            workerIdle = false;
            stateChanged.signalAll();
            if (thread == null || !thread.isAlive()) {
                return true;
            }
            long deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeoutMs);
            while (!workerIdle) {
                long remaining = deadline - System.nanoTime();
                if (remaining <= 0L) {
                    return false;
                }
                try {
                    idleSignal.awaitNanos(remaining);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return false;
                }
            }
            return true;
        } finally {
            lock.unlock();
        }
    }

    public boolean pause() {
        return pause(2000L);
    }

    public void resume() {
        lock.lock();
        try {
            pauseRequested = false;
            workerIdle = false;
            stateChanged.signalAll();
        } finally {
            lock.unlock();
        }
    }

    private void run() {
        while (!stopRequested) {
            if (pauseRequested) {
                lock.lock();
                try {
                    // Park until resume/stop. Re-announce "I am idle" on
                    // every iteration so a *new* pause() that arrives
                    // while we are still parked from the previous cycle
                    // gets acknowledged immediately, not after the
                    // caller's full pause-timeout.
                    while (pauseRequested && !stopRequested) {
                        workerIdle = true;
                        idleSignal.signalAll();
                        try {
                            stateChanged.await();
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                            stopRequested = true;
                            break;
                        }
                    }
                    workerIdle = false;
                } finally {
                    lock.unlock();
                }
                continue;
            }

            Map<String, Object> change;
            try {
                change = primary.nextChange(pollTimeoutMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
            if (change == null) {
                continue;
            }
            try {
                cache.applyChange(change);
            } catch (Exception exc) {
                // Demo behaviour: log and drop the event. A production CDC
                // consumer would retry with bounded backoff and expose a
                // dead-letter / error counter; see the guide's "Production
                // usage" section.
                System.err.println("[sync] failed to apply " + change + ": " + exc);
            }
        }
    }
}
